### ISSUE TITLE
sql: better batching (think 1pc txns)

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#3417]
With `CREATE TABLE bench.insert (k INT PRIMARY KEY)`, the following (and pretty much everything else) will run two KV batches, one for the insert and a separate `EndTransaction`:

`INSERT INTO bench.insert VALUES (1)`

We need to optimize the SQL server to batch up as many commands as it
can (basically until it reads data) and commit in the same round-trip
if possible. Currently the Txn object is used deep down in the individual
operations (for example (*planner).Insert()) so this isn't trivial, at least not
in general.

This will show up prominently in any sort of benchmark, but in particular when
network latency becomes noticeable.

I suggest we make this a high-priority item on the performance shortlist (at least
for simple inserts).

**Comment by petermattis** [comment]
SQL operations currently use `planner.txn` to perform KV oeprations, which is a `*client.Txn`. I think we could change this to a `sql.internalTxn` structure which would provide the same interface as `client.Txn` (and internally contain a `client.Txn`), but would also contain a `client.Batch`. Write operations would be  directed to the batch. Read operations would flush the batch and then be directed to the txn. I haven't fully thought through the details here, but it seems this could be done with minimal impact on the existing SQL operation implementations.

**Comment by tschottdorf** [comment]
Sounds good. Error handling might be a little tricky (you get the error in a different place, so possibly some logic has to be attached to the "flush"). If this `internalTxn` idea pans out, it's also going to be a good way of counting "effective" round trips at that level (the coordinator will incur more round-trips internally, but this is going to be an interesting high-level number to optimize).

**Comment by petermattis** [comment]
Good point about error handling. If a user performs a series of `INSERT` operations within a transaction, do we have to return an error on a unique index violation as soon as it occurs? Or can we delay the error to the first read operation or commit?

**Comment by bdarnell** [comment]
It's a common practice for people to do something like this:

```
try {
   doInsert(txn);
} catch (UniqueConstraintViolation) {
  doUpdate(txn)
}
```

So people do expect the error to be raised at the point where the insert was performed.

**Comment by petermattis** [comment]
That's what I was afraid of. Perhaps the behavior could be controlled by a flag in the session. We could still optimize auto-txns (i.e. statements executed outside of an explicit transaction) and multiple-statements sent in a single request within a transaction.

**Comment by bdarnell** [comment]
PostgreSQL has the `SET CONSTRAINTS DEFERRED` command to delay certain constraint checks to the end of the transaction (it only applies to foreign key constraints but we could adopt the same syntax here). Or we could add a modifier to the `INSERT` command, but there doesn't seem to be anything standard we could adopt (older versions of mysql had `INSERT DELAYED` which meant to return before the insert was applied).

Is `INSERT` the only command that could be optimized as write-only? I.e. is there any case in which `UPDATE` or `DELETE` would not involve a combination of reads and writes?

**Comment by tschottdorf** [comment]
> It's a common practice for people to do something like this:

I assumed we'd always flush the batch (at least for now) before returning to the client. Simple point updates, inserts and deletes should all be possible on one internal round-trip, end-to-end. If you send five requests to the coordinator, it seems very reasonable to expect the coordinator to also run five requests against the cluster, so while we could add a new feature there I don't see it as a must-have for now (especially since it's not supported by vanilla SQL).

**Comment by petermattis** [comment]
Inserts are possible in a single round-trip (assuming no secondary indexes). Point updates and deletes are more complex. Currently updates and deletes are implemented as a select (to find the rows to update/delete) followed by the update/delete of the appropriate data. We'd have to recognize when the where-clause fully and solely constrained the rows to update/delete. It could certainly be done, but it wouldn't be at all pleasant with the current query planner/execution framework.

**Comment by tschottdorf** [comment]
Let's go for single-roundtrip inserts for now. Even with secondary index updates, we should be able to get one round trip, correct? I'm talking about round-trips caused by `client.Txn`. Of course, there will be more internally unless everything we touch is colocated.

**Comment by petermattis** [comment]
As always, it is good to experiment before proceeding in with a significant change. The "old" numbers are from master (as of this morning). The "new" numbers are for performing all of the writes on an INSERT operation within a batch and then committing the batch. This work was a bit of a hack. Error handling is not performed properly and it looks like getting it right will be a significant headache.

As might be expected, the numbers show decreasing benefit as we insert more rows per operation. It looks like we should definitely push forward with this optimization.

```
name                   old time/op    new time/op    delta
Insert1_Cockroach-8       941µs ± 0%     585µs ± 2%  -37.82%   (p=0.000 n=8+10)
Insert10_Cockroach-8     1.31ms ± 1%    0.96ms ± 0%  -26.80%   (p=0.000 n=9+10)
Insert100_Cockroach-8    4.76ms ± 1%    4.42ms ± 1%   -7.24%  (p=0.000 n=10+10)

name                   old alloc/op   new alloc/op   delta
Insert1_Cockroach-8       100kB ± 1%      58kB ± 0%  -41.55%    (p=0.000 n=9+8)
Insert10_Cockroach-8      169kB ± 0%     123kB ± 0%  -26.86%   (p=0.000 n=9+10)
Insert100_Cockroach-8     817kB ± 0%     771kB ± 0%   -5.65%   (p=0.000 n=9+10)

name                   old allocs/op  new allocs/op  delta
Insert1_Cockroach-8       1.31k ± 0%     0.83k ± 0%  -36.65%   (p=0.000 n=9+10)
Insert10_Cockroach-8      2.14k ± 0%     1.64k ± 0%  -23.30%  (p=0.000 n=10+10)
Insert100_Cockroach-8     10.0k ± 0%      9.5k ± 0%   -5.67%   (p=0.000 n=10+7)
```

**Comment by tschottdorf** [comment]
no surprise there, that's essentially what the corresponding kv-type
benchmark showed when I introduced 1pc txns. It'll obviously be way worse
with actual latency.

**Comment by andreimatei** [comment]
I'd probably be interested in working on this. Sounds like a nice random
walk through the code.

On Sun, Dec 13, 2015 at 9:37 PM, Tobias Schottdorf <notifications@github.com

**Comment by petermattis** [comment]
Happy to hand it off. You can see the branch I created [here](https://github.com/cockroachdb/cockroach/tree/pmattis/sql-batched-txn). We should probably talk through whether the existing approach is the right one or if something more tailored to INSERT would simplify the error handling issue.

**Comment by andreimatei** [comment]
Oh I didn't realize you're already on it. So... whatever you want to do.
I'm still probably tied for a day or two on the indexes stuff.

**Comment by petermattis** [comment]
Let's circle back when you're done with the indexes stuff. I doubt I'll get to this in the next day or two.

**Comment by bdarnell** [comment]
Something to think about as we're designing this: certain `UPDATE` patterns can be optimized in special ways: `UPDATE tbl SET x=x+1 WHERE pk=$1` can be turned into a one-phase transaction using the KV increment request. We probably won't get to this kind of optimization any time soon, but it's a non-`INSERT` use case to consider if it doesn't complicate the work in the short term.

**Comment by petermattis** [comment]
We'd have to be sure that `x` is not used in an index. If we were simply assigning a value to `x` we could also optimize this case if we had a version of `CPut` that simply ensured a previous value existed.

**Comment by tamird** [comment]
@petermattis can you summarize the remaining work here?

**Comment by petermattis** [comment]
I'm not sure if there is anything remaining to be done. @bdarnell was suggesting that we could optimize `UPDATE tbl SET x=x+1 WHERE pk=$1` to use a KV increment request, but that is going to be problematic with column families. @paperstreet I believe we optimize `UPDATE tbl SET x=1 WHERE pk=$1` in simple kv-style situations, correct?

**Comment by paperstreet** [comment]
@petermattis Which optimization are you thinking of? UPDATE supports autoCommit, so will send the endtransaction with the puts, but I'm fairly sure it always fetches before computing the mutations.

**Comment by petermattis** [comment]
Perhaps I was thinking of `UPSERT`. Isn't there a code path where we don't fetch first?

**Comment by paperstreet** [comment]
Ah, yeah, I should have realized that's what you were thinking of. If you UPSERT (or a very special case of INSERT/ON CONFLICT) and don't have any indexes, it doesn't fetch first.

**Comment by petermattis** [comment]
Ok. I'm closing this from now. Using the KV increment op is no longer feasible given column families. Pushing more SQL smarts down to the KV layer is a much bigger project.
